Refactor Spack's URL parsing commands

adamjstewart opened the pull request:
Closes #2926

This PR combines the following commands:
```
spack url-parse
spack urls
```
into a single command with subcommands:
```
spack url parse
spack url list
spack url test
```

### What are these commands and why do I care?

These commands aren't commonly used by Spack users, but are absolutely essential for developers to debug Spack's URL parsing support. This includes automatic name detection in `spack create` and version extraction and substitution in commands like `spack fetch`, `spack versions`, `spack checksum`, `spack info`, and `spack create`. Let's look at some examples:

### spack url parse

You are developing a package but you noticed that the version wasn't detected properly. Let's look at #2840 as an example. To determine what is going wrong, you would run:
```
$ spack url parse https://github.com/tpaviot/oce/archive/OCE-0.18.tar.gz
==> Parsing URL: https://github.com/tpaviot/oce/archive/OCE-0.18.tar.gz

==> Matched version regex  5: r'github.com/[^/]+/[^/]+/archive/(?:release-)?v?(\\w+(?:[.-]\\w+)*)$'
==> Matched  name   regex  4: r'github.com/[^/]+/([^/]+)/archive'

==> Detected:
    https://github.com/tpaviot/oce/archive/OCE-0.18.tar.gz
                               ---         ~~~~~~~~
    name:    oce
    version: OCE-0.18

==> Substituting version 9.9.9b:
    https://github.com/tpaviot/oce/archive/OCE-0.18.tar.gz
                               ---         ~~~~~~~~
```
This command tells you which regexes were used to determine the name of the package and the version. It also shows you what it would look like if we substituted a different version into this URL, as is done every time you fetch something. If given the `--spider` option, it spiders the web for more versions.

As you can see, we did a terrible job matching OCE. I plan on fixing this and several other low-hanging fruit in a separate PR once this is merged.

### spack url list

This command lists every URL in every package in Spack. If given the `--color` and `--extrapolation` flags, it also colors the part of the string that it detected to be the name and version. I added `--incorrect-name` and `--incorrect-version` flags to only print URLs that were not being parsed correctly. This is how one identifies said low-hanging fruit.

### spack url test

This command prints a summary that tells us how good of a job we are doing. It also prints the regular expressions that were matched and how commonly they were matched. I know you're all curious to see what kind of a report card we are getting, so here it is:
```
$ spack url test
==> Generating a summary of URL parsing in Spack...

    Total URLs found:          1145
    Names correctly parsed:    1022/1145 (89.26%)
    Versions correctly parsed: 1049/1145 (91.62%)

==> Statistics on name regular expresions:

    Index  Count  Regular Expresion
      1:     11   r'github.com/[^/]+/[^/]+/releases/download/2.2/(.*)-2.2$'
      2:     16   r'/([^/]+)/(tarball|zipball)/'
      3:      6   r'/([^/]+)[_.-](bin|dist|stable|src|sources)[_.-]0'
      4:    112   r'github.com/[^/]+/([^/]+)/archive'
      5:      1   r'[^/]+/([^/]+)/repository/archive'
      6:      4   r'([^/]+)/download.php'
      7:    953   r'([^/]+)[_.-]v?5.1.1'
      8:     36   r'([^/]+)2.10.1'
     10:      1   r'/([^/]+)0.20.1'

==> Statistics on version regular expresions:

    Index  Count  Regular Expresion
      0:      8   r'github.com/.+/(?:zip|tar)ball/v?((\\d+\\.)+\\d+)$'
      1:      1   r'github.com/.+/(?:zip|tar)ball/.*-((\\d+\\.)+\\d+)$'
      4:      1   r'github.com/[^/]+/yorick/archive/y_(\\d+(?:_\\d+)*)$'
      5:    110   r'github.com/[^/]+/[^/]+/archive/(?:release-)?v?(\\w+(?:[.-]\\w+)*)$'
      7:     34   r'github.com/[^/]+/[^/]+/releases/download/v?([^/]+)/.*$'
      8:      1   r'\\?ref=(?:.*-|v)*((\\d+\\.)+\\d+).*$'
     10:      3   r'((\\d+_)+\\d+)$'
     11:      5   r'-((\\d+\\.)*\\d\\.\\d+-(p|rc|RC)?\\d+)(?:[-._](?:bin|dist|stable|src|sources))?$'
     12:      2   r'-((\\d)+-\\d)'
     13:    184   r'_((\\d+\\.)+\\d+(-(\\d+(\\.\\d+)?))?[a-z]?)'
     14:    683   r'-((\\d+\\.)*\\d+)$'
     15:     21   r'[-._]?v?((\\d+\\.)*\\d+[-._]?([a-z]|rc|RC|tp|TP?)\\d*)$'
     17:     28   r'((\\d+\\.)*\\d+)$'
     18:     17   r'-((\\d+\\.)+\\d+[a-z]?)[-._](bin|dist|stable|src|sources?)$'
     20:     20   r'-v?([^-]+(-alpha|-beta)?)'
     21:      7   r'_([^_]+(_alpha|_beta)?)'

```
Overall, not as bad as I thought. But there is still a lot that can be done to improve this. With these new debugging capabilities, I can safely and confidently add or remove regular expressions from the list and immediately see which packages were improved or broken.

adamjstewart commented:
@tgamblin I'll check it out. I also plan on adding documentation for `spack url parse` in the Packaging Guide and `spack url list/test` in the Developer Guide. I think there's at least one documentation request I can knock off relating to `list_url` too.

adamjstewart commented:
So the helper functions like `name_parsed_correctly` would be easy to test, but the rest of the commands just print output that will change every time a url or regex changes or a package is added. I'm not sure what you want me to test, other than that the commands complete without crashing.

adamjstewart commented on lib/spack/spack/test/cmd/url.py:
```diff
@@ -0,0 +1,69 @@
+
+# def test_url_parse(parser):
+#     args = parser.parse_args(['parse', 'http://zlib.net/fossils/zlib-1.2.10.tar.gz'])
+#     spack.cmd.url.url(parser, args)
```
I wasn't able to get this working. I want to add basic tests just to confirm that `spack url list` and `spack url test` don't crash.

adamjstewart commented on lib/spack/spack/test/cmd/url.py:
```diff
@@ -0,0 +1,69 @@
+    assert name_parsed_correctly(Package('netcdf',         []), 'netcdf')
+    assert name_parsed_correctly(Package('r-devtools',     []), 'devtools')
+    assert name_parsed_correctly(Package('py-numpy',       []), 'numpy')
+    assert name_parsed_correctly(Package('octave-splines', []), 'splines')
```
I was trying to figure out how to use actual package classes, but I couldn't figure it out.

adamjstewart commented on lib/spack/spack/test/cmd/url.py:
```diff
@@ -0,0 +1,69 @@
+#     """Returns the parser for the ``url`` command"""
+#     parser = argparse.ArgumentParser()
+#     setup_parser(parser)
+#     return parser
```
I've never used pytest before, so please give me pointers if you have any.

alalazo replied:
`scope` can be one of `session`, `module`, `class`, `function`. You can get a `parser` in your tests just using the following naming convention (known as "funcargs" in pytest context):
```python
def test_foo(parser):
    # You get a parser object constructed as ``parser = parser()``
    ....
```
See [here](http://pytest.org/dev/fixture.html) for the complete documentation on fixtures.

adamjstewart replied:
Ah, I just copied this from the `module.py` test, which called `spack module` and had a scope of `module`, so I assumed it was the name of the command that I had to use.

adamjstewart commented:
I see we switched to codecov. How do I check how much my unit tests change our coverage?

adamjstewart commented:
Sweet, tests look pretty good. I didn't test `spack url parse --spider` because I wasn't sure whether or not we wanted tests that require internet access. I also need to figure out how to add expected failures for `spack url parse`.

tgamblin commented:
Nice!  It's supposed to be possible to get codecov to add status checks and not just comment (I think the comments are kind of spammy).  I'll see if I can figure that out, but the report is pretty informative.

adamjstewart commented:
@tgamblin When does it decide to leave a comment? I agree that status checks would take up less of the feed.

adamjstewart commented:
Added an expected failure for `spack url parse`. Should I add a test for `spack url parse --spider`? I wasn't sure whether or not tests should require internet access. As far as I know, that's the only part of the code that isn't currently covered.

adamjstewart commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -728,6 +728,102 @@ versions. See the documentation on `attribute_list_url`_ and
 
+--------------------
+Finding new versions
+--------------------
```
I moved this section up to be directly below "Versions and fetching" instead of down below "Handling RPATHs". I think it makes much more sense for all of the version/URL stuff to be together.

adamjstewart commented:
Ok, I added documentation on `spack url` and adding new commands, I added docstrings to all of the methods I modified, and I added unit tests that should cover every line of `spack url` except for `spack url parse --spider`. Let me know if there's anything else you want to see in this PR. If not, I'll take a swing at overhauling our regexes to match more URLs.

adamjstewart commented:
Looks like the tests are failing because commands can't have non-integer return values. I think I can work around that.

tgamblin commented:
@adamjstewart: this looks great to me.  Thanks for the awesome refactor!
